mantidproject/mantid
# Monitor for external changes to script file

**Description of work**
When another application modifies a script that's open in Workbench, then the user will now get a notification. They can either reload the file, or keep their version open but unsaved. This is the same behaviour as Notepad++. To do this there are multiple components, monitoring the files (but not when it's Workbench doing the saving), reloading the file, and marking the file as
modified.

**Purpose of work**

Fixes #34981 



**Summary of work**

- Monitor scripts for changes with a QFileSystemWatcher
- When saving the scripts from Workbench, temporarily disable the file watcher
- On focus (required some signal plumbing), prompt the user if the script has changed externally. They can either reload the file (new method), or they can mark the file as modified (new method).
- Cppcheck suppression added for the new signal on ScriptEditor. It looks like trying to emit a signal from C++ will give you a shadow function error in Cppcheck, at least in the version we're currently using.


**To test:**
Open some scripts in Workbench and another editor. Make changes in either Workbench or the other editor and see how Workbench handles the situation. There are two options when the script has changed externally, either reload the script or keep the Workbench version but marked as modified, would be good to test both those options are working correctly. Can also try deleting or renaming a script file that's open in Workbench.







---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
